### PR TITLE
[GEOS-11225] AuthKey synchronize the user/group automatically

### PR DESCRIPTION
**Pull request by afabiani**
[GEOS-11225](https://osgeo-org.atlassian.net/browse/GEOS-11225) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

As per [GEOS-11225](https://osgeo-org.atlassian.net/browse/GEOS-11225) the idea of this pull request would be to add an option to the `AuthKey Provider Mappers` to eventually be synchornized automatically.

image

The proposed implementation would be to introduce a sheduler task which checks the `AuthKey` provider configurations, checks if the `auto-sync` option is enabled, and if it is the case, invoke the `Sync` method on it.

_Open Questions_ How to deal with `Web Services` based auth-keys? Should we store the `auth-keys` form the web-services somewhere?

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/extension/authkey/src/main/java/org/geoserver/security/GeoServerAuthenticationKeyProvider.java [hunk 1]
```diff
@@ -7,24 +7,73 @@
...
+        this.securityManager = securityManager;
+        this.autoSyncDelaySeconds = autoSyncDelaySeconds;
+        this.scheduler = Executors.newScheduledThreadPool(1, getThreadFactory());
+
+        // schedule auto-sync thread
+        scheduler.scheduleAtFixedRate(
+                authKayMapperSyncTask,
+                autoSyncDelaySeconds,
+                autoSyncDelaySeconds,
+                TimeUnit.SECONDS);
```
Comment (aaime) [inline src/extension/authkey/src/main/java/org/geoserver/security/GeoServerAuthenticationKeyProvider.java:70]: Shouldn't the delay be user controllable?

File: src/extension/authkey/src/test/java/org/geoserver/security/AuthKeyAuthenticationTest.java [hunk 2]
```diff
@@ -655,6 +655,38 @@ public void testAuthKeyMapperSynchronize() throws Exception {
...
+        config.setAuthKeyMapperName("propertyMapper");
+
+        getSecurityManager().saveFilter(config);
+
+        // File Property Mapper
+        File authKeyFile = new File(getSecurityManager().userGroup().dir(), "testAuthKey");
+        authKeyFile = new File(authKeyFile, "authkeys.properties");
+
+        GeoServerAuthenticationKeyProvider geoServerAuthenticationKeyProvider =
+                new GeoServerAuthenticationKeyProvider(getSecurityManager(), 10);
```
Comment (aaime) [inline src/extension/authkey/src/test/java/org/geoserver/security/AuthKeyAuthenticationTest.java:678]: We normally try to have tests run as fast as possible, maybe make that 2 seconds?

File: src/extension/authkey/src/test/java/org/geoserver/security/AuthKeyAuthenticationTest.java [hunk 3]
```diff
@@ -655,6 +655,38 @@ public void testAuthKeyMapperSynchronize() throws Exception {
...
+        assertNotNull(geoServerAuthenticationKeyProvider.getScheduler());
+        assertFalse(geoServerAuthenticationKeyProvider.getScheduler().isTerminated());
+        assertFalse(geoServerAuthenticationKeyProvider.getScheduler().isShutdown());
+        assertEquals(10, geoServerAuthenticationKeyProvider.getAutoSyncDelaySeconds());
+        Thread.sleep(2 * geoServerAuthenticationKeyProvider.getAutoSyncDelaySeconds() * 1000L);
```
Comment (aaime) [inline src/extension/authkey/src/test/java/org/geoserver/security/AuthKeyAuthenticationTest.java:683]: Due to runtime issues and Github actions VM unpredictability, waiting exactly the time configured might not be enough. Check the code for usage of Awaitility, it helps set up a wait for a given condition, and with a max timeout that allows for some tolerance to slow running environments.

**Review by aaime — CHANGES_REQUESTED** [review]
For keys received from web services... is there any sort of caching going on? Why would you need to store the results?

**Comment by afabiani** [comment]
> For keys received from web services... is there any sort of caching going on? Why would you need to store the results?

Actually the requirement is to purge the cache on synchronize in order to be sure the stored permissions are aligned with the remote service ones.

File: src/extension/authkey/src/main/java/org/geoserver/security/GeoServerAuthenticationKeyProvider.java [hunk 4]
```diff
@@ -75,4 +124,79 @@ public boolean roleServiceNeedsLockProtection() {
...
+    public ScheduledExecutorService getScheduler() {
+        return scheduler;
+    }
+
+    public int getAutoSyncDelaySeconds() {
+        return autoSyncDelaySeconds;
+    }
+
+    /** Runnable task to synchronize the AuthenticationKeyMapper */
+    private final Runnable authKayMapperSyncTask =
```
[the remark this answers is not in the document]
Comment (afabiani) [reply src/extension/authkey/src/main/java/org/geoserver/security/GeoServerAuthenticationKeyProvider.java]: Done, thanks for the hint.

File: src/extension/authkey/src/test/java/org/geoserver/security/AuthKeyAuthenticationTest.java (hunk 2 above)
Comment (afabiani) [reply src/extension/authkey/src/test/java/org/geoserver/security/AuthKeyAuthenticationTest.java]: Done, thanks for the hint.

File: src/extension/authkey/src/test/java/org/geoserver/security/AuthKeyAuthenticationTest.java (hunk 3 above)
Comment (afabiani) [reply src/extension/authkey/src/test/java/org/geoserver/security/AuthKeyAuthenticationTest.java]: Done, thanks for the hint.

File: src/extension/authkey/src/main/java/org/geoserver/security/GeoServerAuthenticationKeyProvider.java (hunk 1 above)
Comment (afabiani) [reply src/extension/authkey/src/main/java/org/geoserver/security/GeoServerAuthenticationKeyProvider.java]: The mappers are prototypes and the schedulers will be removed from memory within the filter configuration. Moreover the delay seconds can be eventually tweaked from the applicationContext and this is more a DevOps tuning configuration other than a user one IMHO.

**Comment by afabiani** [comment]
Build jobs are failing due to en error with `[ERROR] Failed to execute goal on project gs-grib: Could not resolve dependencies for project org.geoserver.extension:gs-grib:jar:2.25-SNAPSHOT`. Not sure how it is related to this PR.

**Comment by aaime** [comment]
The issue with repositories should have been fixed yesterday. Closing and reopening to force a new set of builds.

**Review by aaime — APPROVED** [review]

**Comment by aaime** [comment]
Changes look good, one thing, I noticed the new UI bit is not documented (part of the PR checklist).

**Comment by afabiani** [comment]
Thanks @aaime. Documentation updated too.

**Comment by aaime** [comment]
Merged, thanks!

**Comment by aaime** [comment]
The backport to `2.23.x` failed:
```
The process '/usr/bin/git' failed with exit code 128
```
<details>
<summary>stderr</summary>

```
error: commit fc6e34377dcfbdff46ed044f7a589590b36614e7 is a merge but no -m option was given.
fatal: cherry-pick failed

```
</details>

<details>
<summary>stdout</summary>

```
[backport-7316-to-2.23.x c054c09ffc] [GEOS-11225] AuthKey synchronize the user/group automatically
 Author: afabiani <alessio.fabiani@geosolutionsgroup.com>
 Date: Thu Dec 7 15:54:38 2023 +0100
 7 files changed, 187 insertions(+), 11 deletions(-)

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-2.23.x 2.23.x
# Navigate to the new working tree
cd .worktrees/backport-2.23.x
# Create a new branch
git switch --create backport-7316-to-2.23.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick ba2df99a70f693737fbdd7c93f0bfb3f141f315a,fc6e34377dcfbdff46ed044f7a589590b36614e7,f61f37528f1f1fed2706236073410547e530e2fe,81c538124260e429844605016b3038b901e76e8a,e4e5f3ad0b0a893172f5586abdf948b6e59aaf17,f3e4dacb4db6bcfb2c7cfd77c421960a704865d7,53a2e8fc6f31c3b2a5fb93994d6289e969dff80b,2a0f31144848ba01769e7f5d31a3342e7521cd45,19a596e92cb8e26d44a5f678154c4c6c17fc7caf,e9604b9c965b860ae80fa83f09e6c449291f1756
# Push it to GitHub
git push --set-upstream origin backport-7316-to-2.23.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-2.23.x
```
Then, create a pull request where the `base` branch is `2.23.x` and the `compare`/`head` branch is `backport-7316-to-2.23.x`.

**Comment by aaime** [comment]
The backport to `2.24.x` failed:
```
The process '/usr/bin/git' failed with exit code 128
```
<details>
<summary>stderr</summary>

```
error: commit fc6e34377dcfbdff46ed044f7a589590b36614e7 is a merge but no -m option was given.
fatal: cherry-pick failed

```
</details>

<details>
<summary>stdout</summary>

```
Auto-merging src/extension/authkey/src/test/java/org/geoserver/security/AuthKeyAuthenticationTest.java
[backport-7316-to-2.24.x c685328c4e] [GEOS-11225] AuthKey synchronize the user/group automatically
 Author: afabiani <alessio.fabiani@geosolutionsgroup.com>
 Date: Thu Dec 7 15:54:38 2023 +0100
 7 files changed, 187 insertions(+), 11 deletions(-)

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-2.24.x 2.24.x
# Navigate to the new working tree
cd .worktrees/backport-2.24.x
# Create a new branch
git switch --create backport-7316-to-2.24.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick ba2df99a70f693737fbdd7c93f0bfb3f141f315a,fc6e34377dcfbdff46ed044f7a589590b36614e7,f61f37528f1f1fed2706236073410547e530e2fe,81c538124260e429844605016b3038b901e76e8a,e4e5f3ad0b0a893172f5586abdf948b6e59aaf17,f3e4dacb4db6bcfb2c7cfd77c421960a704865d7,53a2e8fc6f31c3b2a5fb93994d6289e969dff80b,2a0f31144848ba01769e7f5d31a3342e7521cd45,19a596e92cb8e26d44a5f678154c4c6c17fc7caf,e9604b9c965b860ae80fa83f09e6c449291f1756
# Push it to GitHub
git push --set-upstream origin backport-7316-to-2.24.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-2.24.x
```
Then, create a pull request where the `base` branch is `2.24.x` and the `compare`/`head` branch is `backport-7316-to-2.24.x`.

**Pull request merged**